index.rst

```diff
@@ -1,61 +1,461 @@
+
+.. _data types: https://docs.influxdata.com/influxdb/v1.8/write_protocols/line_protocol_reference/#data-types
+.. _long debate: https://github.com/influxdata/influxdb/issues/4089
+
+
+We can show that by running a local instance of InfluxDB.
+
+.. jupyter-execute::
+
+  %%bash
+  docker-compose up -d influxdb
+
+By default, the InfluxDB API runs at http://localhost:8086. Let's use the Python ``requests`` module to interact with the InfluxDB API and create the ``mydb`` database for our tests:
+
+.. jupyter-execute::
+
+  import requests
+  requests.post(url="http://localhost:8086/query", params={'q':'CREATE DATABASE "mydb"'})
+
+
+Now let's write a point to the measurement ``foo`` with a field key ``bar`` and value ``NaN``:
+
+
+.. jupyter-execute::
+
+  p = "foo bar=NaN"
+  r = requests.post(url="http://localhost:8086/write?db=mydb", data=p)
+  r.text
+
+This confirms that we cannot use ``NaN`` in InfluxDB to represent a missing value for the ``float`` type.
+
+However, because InfluxDB is a schema-less database, we can change the schema on write.
+We can add new fields or drop existing fields at any time.
+This suggests that fields are optional and that InfluxDB should automatically fill them in with a default value.
+
+To verify this property of InfluxDB, let's write a sequence of points, and change the schema as we write.
+
+.. jupyter-execute::
+  :emphasize-lines: 2
+
+  p1 = "foo bar=1.0,baz=1.0"
+  p2 = "foo baz=2.0"
+  p3 = "foo bar=3.0,baz=3.0"
+  requests.post(url="http://localhost:8086/write?db=mydb", data=p1)
+  requests.post(url="http://localhost:8086/write?db=mydb", data=p2)
+  requests.post(url="http://localhost:8086/write?db=mydb", data=p3)
+
+The following query returns the ``foo`` measurement:
+
+.. jupyter-execute::
+
+  r = requests.get(url="http://localhost:8086/query", params={'q': 'SELECT * FROM "mydb"."autogen"."foo"'})
+  r.json()['results'][0]['series']
+
+Notice that when querying the foo measurement, InfluxDB returns ``None`` for the missing value of bar in the second point.
+The Python keyword ``None`` is used here as the default value for an optional parameter (the InfluxDB field in this case), as expected.
+
+The InfluxDB Sink connector is the right place for dropping missing values. In particular, the ``influxdb-java`` library used by the connector, `skips fields with NaN values`_ when writing to InfluxDB.
+This feature was implemented upstream a month after we first notice this problem back in September 2019, during the AuxTel integration activities at the Summit.
+
+.. _skips fields with NaN values: https://github.com/influxdata/influxdb-java/blob/master/CHANGELOG.md#216-2019-10-25
+.. _writing NaN values to InfluxDB: https://jira.lsstcorp.org/browse/DM-21300
+
+
+In summary, fields are optional, and we should drop missing values when writing to InfluxDB.
```

Comment:
Maybe you are going to cover this later on, but this still doesn't address the issue that we don't know if a field is missing because it was never filled in or because it was intentionally set to NaN.

Reply:
It is just designed as missing in the EFD. I don't think we can distinguish these two situations.  I'll clarify this further in the introduction when we describe the situations when missing values can occur.